### ISSUE TITLE
FR: native handling of MappingProxyType

### ISSUE THREAD

**Issue by ofek** [pydantic/pydantic#6868]
https://docs.python.org/3/library/types.html#types.MappingProxyType

Currently in order to use that one must set https://docs.pydantic.dev/dev-v2/usage/model_config/#arbitrary-types-allowed

Selected Assignee: @davidhewitt

Selected Assignee: @hramezani

**Comment by davidhewitt** [comment]
> You could probably implement this all in the `pydantic` side if you wanted to. I don't think (not sure) that there's a C API for this stuff so it's unlikely to be orders of magnitude faster in Rust or anything like that. Let me know if you want pointers on how you'd do it in Python.

There actually is a C API and a PyO3 PR (unfinished) to support `mappingproxy` at https://github.com/PyO3/pyo3/pull/2435

Given I assume we will just call `types.MappingProxyType(validated_dict)` as a post-processing step, the amount of effort to move this to Rust may not be worth the minimal perf benefit. If someone was really interested in writing this in Rust I wouldn't stop them of course 😉

**Comment by dmontagu** [comment]
If the goal is to convert _all_ dicts to `mappingproxy`s, I'll just note that the recent changes to allow a custom `GenerateSchema` might be sufficient? Not sure if we have exposed the ability to override `dict_schema` or just `str_schema` at this point.

**Comment by ofek** [comment]
My PR in the other repository inadvertently closed this, can you please re-open?

**Comment by samuelcolvin** [comment]
done.

**Comment by samuelcolvin** [comment]
> If the goal is to convert _all_ dicts to `mappingproxy`s, I'll just note that the recent changes to allow a custom `GenerateSchema` might be sufficient? Not sure if we have exposed the ability to override `dict_schema` or just `str_schema` at this point.

This didn't get merged in the end, we're still working on how to make it possible.

**Comment by bepuca** [comment]
Hello! I wonder what is the state of this? I also saw this PR was merged and it is unclear if changes anything: https://github.com/PyO3/pyo3/pull/4644

**Comment by davidhewitt** [comment]
What is the use case you have? Yes PyO3 has bindings for `MappingProxyType`, so a PR to support is probably possible.

**Comment by bepuca** [comment]
My use case is that I am defining my data model through Pydantic models. One of this data items, contains a field that would be ideally of type `dict[str, Something]`, because these Somethings have id and I need to retrieve them by id. At the same time, I want the models to be frozen. They are the in memory representation of data, but they should not be mutable. The problem with using a regular dict is that this immutability is broken because I can change the value of the key and Pydantic will not raise (nor the structure).

I worked around that through validators and serialization so on the outside I have this dict but once the model exists is actually a frozenset. But then I need a getter that iterates through my tuple key value pairs to find the right key. Not the end of the world, but it does add some bloat and I was wondering if this would work.

So basically the use case where I see this is: I need a dict in a model that should be immutable. Maybe there are other ways to solve this more nicely. If so, happy to hear them!
